spec/requests/analytics/churn_spec.rb

```diff
@@ -0,0 +1,129 @@
+      churned_sub2 = create(:subscription, link: yearly_product, user: create(:user),
+                                           created_at: "2023-12-01 12:00:00", deactivated_at: "2023-12-25 12:00:00")
+
+      [active_sub1, active_sub2, new_sub, churned_sub1].each do |sub|
+        payment_option = create(:payment_option, subscription: sub, price: monthly_price)
+        sub.update!(last_payment_option: payment_option)
+      end
+
+      yearly_payment = create(:payment_option, subscription: churned_sub2, price: yearly_price)
+      churned_sub2.update!(last_payment_option: yearly_payment)
+    end
+
+    it "calculates total stats" do
+      visit churn_dashboard_path(from: "2023-12-01", to: "2023-12-31")
+
+      within_section("Churn rate") { expect(page).to have_text("40.0%") }
+      within_section("Last period churn rate") { expect(page).to have_text("0.0%") }
+      within_section("Revenue lost") { expect(page).to have_text("$20") }
+      within_section("Churned users") { expect(page).to have_text("2") }
+    end
+
+    it "allows filtering by product" do
+      visit churn_dashboard_path(from: "2023-12-01", to: "2023-12-31")
+
+      within_section("Churn rate") { expect(page).to have_text("40.0%") }
+      within_section("Revenue lost") { expect(page).to have_text("$20") }
+      within_section("Churned users") { expect(page).to have_text("2") }
+
+      select_disclosure "Select products..." do
+        uncheck "Annual Plan"
+      end
+
+      within_section("Churn rate") { expect(page).to have_text("25.0%") }
+      within_section("Revenue lost") { expect(page).to have_text("$10") }
+      within_section("Churned users") { expect(page).to have_text("1") }
+
+      select_disclosure "Select products..." do
+        check "Annual Plan"
+      end
+
+      within_section("Churn rate") { expect(page).to have_text("40.0%") }
+      within_section("Revenue lost") { expect(page).to have_text("$20") }
+      within_section("Churned users") { expect(page).to have_text("2") }
+
+      select_disclosure "Select products..." do
+        uncheck "Monthly Membership"
+      end
+
+      within_section("Churn rate") { expect(page).to have_text("100.0%") }
+      within_section("Revenue lost") { expect(page).to have_text("$10") }
+      within_section("Churned users") { expect(page).to have_text("1") }
+    end
+
+    it "allows custom date range filtering" do
+      visit churn_dashboard_path(from: "2023-12-01", to: "2023-12-31")
+
+      within_section("Churned users") { expect(page).to have_text("2") }
+      within_section("Churn rate") { expect(page).to have_text("40.0%") }
+
+      visit churn_dashboard_path(from: "2023-12-20", to: "2023-12-31")
+      within_section("Churned users") { expect(page).to have_text("2") }
+
+      visit churn_dashboard_path(from: "2023-12-01", to: "2023-12-19")
+      within_section("Churned users") { expect(page).to have_text("0") }
+    end
+
+    it "prevents date ranges exceeding 31 days" do
```

Comment:
> it "prevents date ranges exceeding 31 days" do

The designs [here](https://github.com/antiwork/gumroad/issues/13#issuecomment-2969920369) seem to show churn rate by day if weekly or monthly view is chosen, and by month for yearly and all time view I think. Don't see a restriction of 31 days in the issue requirements, so this doesn't seem correct?

Reply:
hey @hchhabra , this requirement of the time window only being a month long was pointed out by @EmCousin, so he can clarify that. But the major gist was that we didn't want to allow calculations going beyond a month since they required way more compute, as that is multiple products and multiple subscribers queried for it